fem/adnonlininteg.hpp

```diff
@@ -0,0 +1,402 @@
+// Copyright (c) 2010, Lawrence Livermore National Security, LLC. Produced at
+// the Lawrence Livermore National Laboratory. LLNL-CODE-443211. All Rights
+// reserved. See file COPYRIGHT for details.
+//
+// This file is part of the MFEM library. For more information and source code
+// availability see http://mfem.org.
+//
+// MFEM is free software; you can redistribute it and/or modify it under the
+// terms of the GNU Lesser General Public License (as published by the Free
+// Software Foundation) version 2.1 dated February 1999.
+
+
+
+#ifndef MFEM_ADNONLININTEG
+#define MFEM_ADNONLININTEG
+
+#include "../config/config.hpp"
+#include "fe.hpp"
+#include "coefficient.hpp"
+#include "fespace.hpp"
+#include "nonlininteg.hpp"
+#include "../linalg/tadvector.hpp"
+#include "../linalg/taddensemat.hpp"
+#include "../linalg/fdual.hpp"
+
+#if defined MFEM_USE_ADEPT
+#include <adept.h>
+#elif defined  MFEM_USE_FADBADPP
+#include <fadiff.h>
+#include <badiff.h>
+#endif
+
+//define Forward AD mode
+//#define MFEM_USE_ADFORWARD
+
+namespace mfem
+{
+
+// m - dimension of the residual vector
+// the Jacobian will have dimensions [m,length(uu)]
+template<template <typename, typename> class CTD, int m>
+class ADQFunctionTJ
+{
+protected:
+#ifdef MFEM_USE_ADEPT
+   adept::Stack  m_stack;
+#endif
+
+public:
+#if defined MFEM_USE_ADEPT
+   typedef adept::adouble             ADFType;
+   typedef TADVector<ADFType>         ADFVector;
+   typedef TADDenseMatrix<ADFType>    ADFDenseMatrix;
+#elif defined MFEM_USE_FADBADPP
+#ifdef MFEM_USE_ADFORWARD
+   typedef fadbad::F<double>         ADFType;
+   typedef TADVector<ADFType>         ADFVector;
+   typedef TADDenseMatrix<ADFType>    ADFDenseMatrix;
+#else
+   typedef fadbad::B<double>         ADFType;
+   typedef TADVector<ADFType>         ADFVector;
+   typedef TADDenseMatrix<ADFType>    ADFDenseMatrix;
+#endif
+#else
+   typedef mfem::ad::FDual<double>    ADFType;
+   typedef TADVector<ADFType>         ADFVector;
+   typedef TADDenseMatrix<ADFType>    ADFDenseMatrix;
+#endif
+
+#ifdef MFEM_USE_ADEPT
+   ADQFunctionTJ():m_stack(false)  {}
+#else
+   ADQFunctionTJ() {}
+#endif
+
+   ~ADQFunctionTJ() {}
+
+   double QFunction(const mfem::Vector& vparam, mfem::Vector& uu)
```

Comment:
I don't think I can review this file until the comments are added.

Reply:
Comments have been added in the code.